[WIP] Modify, Delete Pipeline feature (#30)

bidhan-a opened the pull request:
Hey @michelvocks 

I have started working on this. 

I've added a new tab pane in the Settings page for managing pipelines. In the backend, I've also added a delete handler which basically does the following: (a) Remove binary, (b) Remove from store, and (c) Remove from the active pipelines list. Is this good enough?

Also, how should we handle the modify/rename thing? Would renaming the binary in the pipelines folder and updating the pipeline in the store & the active pipelines list suffice? Or did you have something else in mind?

Please let me know. 

Thanks!

michelvocks commented on handlers/pipeline.go:
```diff
@@ -136,6 +136,50 @@ func PipelineGet(c echo.Context) error {
+		index++
+	}
+
+	if foundPipeline.Name == "" {
```
You can make foundPipeline a pointer. Then you can check here: if foundPipeline == nil

michelvocks commented:
> Also, how should we handle the modify/rename thing? Would renaming the binary in the pipelines folder and updating the pipeline in the store & the active pipelines list suffice? Or did you have something else in mind?

Yes, I think that's enough for now. I was also thinking about changing the source git repository but then you also have to update credentials etc. So for now renaming should be fine! 🤗

Skarlso commented on handlers/pipeline.go:
```diff
@@ -136,6 +136,50 @@ func PipelineGet(c echo.Context) error {
+	}
+
+	// Look up pipeline for the given id
+	var foundPipeline *gaia.Pipeline
```
There is no need for this to be a reference since you are passing in a copy anyways. Let's keep this on the stack with removing the pointer and the address passing.

bidhan-a replied:
I changed it to a reference type as per @michelvocks suggestion.
> You can make foundPipeline a pointer. Then you can check here: if foundPipeline == nil

Thoughts?

Skarlso replied:
You can check if it is initialized by comparing it to an empty struct of the same type. Rather than a pointer which might be stored on the heap.

Skarlso replied:
I guess it doesn't matter much but avoids a nil pointer expedition.

michelvocks replied:
@bidhan-a Sorry. My mistake here. You implementation was actually right. :sweat_smile:

Skarlso commented on handlers/pipeline.go:
```diff
@@ -136,6 +136,50 @@ func PipelineGet(c echo.Context) error {
+	for pipeline := range pipeline.GlobalActivePipelines.Iter() {
+		if pipeline.ID == pipelineID {
+			foundPipeline = &pipeline
+			deletedPipelineIndex = index
```
Add break here once the pipeline you are looking for is found.

bidhan-a replied:
Hey @Skarlso I tried adding break inside `Iter()` but after doing that the process just freezes. I assume this is why `break` isn't used here either. https://github.com/gaia-pipeline/gaia/blob/master/pipeline/pipeline.go#L102

Skarlso commented on handlers/pipeline.go:
```diff
@@ -136,6 +136,50 @@ func PipelineGet(c echo.Context) error {
+// PipelineDelete accepts a pipeline id and deletes it from the
+// store. It also removes the binary inside the pipeline folder.
+func PipelineDelete(c echo.Context) error {
+	pipelineIDStr := c.Param("pipelineid")
```
There is no need for this variable to exist. `c.Param` will return an empty string if the parameter is not found making Atoi fail gracefully. So this can be removed and inlined into Atoi.

bidhan-a replied:
I thought so too initially but after seeing this pattern used in other handler functions, I decided to stick with it. :smile:

Skarlso commented on handlers/pipeline.go:
```diff
@@ -136,6 +136,50 @@ func PipelineGet(c echo.Context) error {
+	}
+
+	if foundPipeline == nil {
+		return c.String(http.StatusNotFound, err.Error())
```
There is a `errPipelineNotFound` error.

Skarlso commented on handlers/pipeline.go:
```diff
@@ -136,6 +136,50 @@ func PipelineGet(c echo.Context) error {
+	// Delete pipeline binary
+	err = pipeline.DeleteBinary(*foundPipeline)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
```
Instead of an internal server error I suggest creating an `errPipelineDelete` error.  Something like, `oops sorry, I couldn't delete the binary, maybe permissions are incorrect? Please fix permissions and try again or delete it by hand.`. Thoughts?

Skarlso commented on handlers/pipeline.go:
```diff
@@ -136,6 +136,50 @@ func PipelineGet(c echo.Context) error {
+	// Delete pipeline from store
+	err = storeService.PipelineDelete(pipelineID)
+	if err != nil {
+		return c.String(http.StatusNotFound, err.Error())
```
This shouldn't be a `StatusNotFound` not? Rather an internal error since it's a db operation.

bidhan-a commented:
Hey guys @michelvocks @Skarlso 

So I've added the implementation for renaming pipelines. Both renaming & deletion seem to be working properly. Please feel free to critique the code and let me know if I should add/remove anything. If all seems to be okay, I'll start writing test cases for the methods which I've added.

Thanks!

michelvocks commented on handlers/pipeline.go:
```diff
@@ -138,29 +138,45 @@ func PipelineGet(c echo.Context) error {
 		}
+
+		// Update pipeline in store
+		err = storeService.PipelinePut(&p)
```
I think it's better to update the name in ``foundPipeline`` and store this object instead of ``p`` which we got from the client. For example, a hacker could potentially change something else than the name in the struct and that will be saved too.

michelvocks commented on handlers/pipeline.go:
```diff
@@ -138,29 +138,45 @@ func PipelineGet(c echo.Context) error {
+		}
+
+		// Update active pipelines
+		pipeline.GlobalActivePipelines.ReplaceByName(foundPipeline.Name, p)
```
Same here. I think we should stick we the old object 😃

michelvocks commented:
Looks really good now. Good job @bidhan-a 🤗 ❤️ 
Let me know when you are ready with your tests.

bidhan-a commented:
Thanks @michelvocks . Will let you know once tests are added. :smiley:

bidhan-a commented:
Hey @michelvocks I've added test cases for all the newly added functions. Let me know if I've missed anything!
